### PR TITLE
Add trie

### PR DESCRIPTION
**Pull request by akashkatir**
Finally!

### CODE REVIEW

File: src/dsap/tree/trie.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        True
+        >>> trie.search("app")
+        False
+        >>> trie.starts_with("app")
+        True
+    """
+
+    def __init__(self):
+        """Initialize an empty Trie with a root node."""
+        self.trie: TrieNode = TrieNode("")
```
Comment (tshur) [inline src/dsap/tree/trie.py]: You can specify the type outside/above this. Then, you don't need to specify it within the __init__ and it serves as good documentation. Also, lets make it a "private" member variable and call it "_root" maybe? You can check other classes for examples.

i.e.,
```python
_root: TrieNode

def __init__(self):
  ...
  self._root = TrieNode("")
  
...
```

File: tests/tree/trie_test.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+from dsap.tree import Trie
+
+
+class TestTrie:
```
Comment (tshur) [inline tests/tree/trie_test.py]: Nice! I wonder if there are other edge cases we might want to add here, but overall this is fine :)

Example edge cases:
- insert / search / starts_with given empty string. What behavior do we want? Is it working?
- non-standard characters (if we want to allow them), e.g., capital/lowercase treated separately, space, symbols (what about a word containing "*" character?).
- what if we insert the same word multiple times?

Imagine yourself a library designer, and you have to make sure your users (developers) are using the data structure correctly!

**Review by tshur — CHANGES_REQUESTED** [review]
Nice job, looks great! Thanks for getting this in. Left some comments but they should be mostly straightforward to resolve!

File: src/dsap/tree/trie.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    Each node represents a single character and contains references to child nodes
+    for subsequent characters. A special "*" child indicates the end of a complete word.
+
+    Attributes:
+        val (str): The character stored in this node.
+        children (dict[str, TrieNode]): Mapping from characters to child nodes.
+    """
+
+    val: str
+    children: dict[str, "TrieNode"]
```
Comment (tshur) [inline src/dsap/tree/trie.py]: Make these "private" members to help protect the class internals. Do this for all class variables everywhere (unless you actually want users to expect that they can access, modify these directly).

ALSO, here and elsewhere, you can just do TrieNode, not "TrieNode" if you add `from __future__ import annotations` at the top! This enables dynamic type checking (fixes recursive types).

_val: str
_children: dict[str, TrieNode]

**Review by tshur — CHANGES_REQUESTED** [review]

**Review by tshur — APPROVED** [review]
Looks good! Thanks and nice work.

P.S. when you switched to private variable _trie, maybe you realized that in tests you were inspecting private member variables. Typically, this is "off limits" as you want to test behaviors and not implementation details. One workaround would be to implement __bool__ dunder to return if the trie has any data. Then, you won't be exposing internals anymore :)
